File: 3493.properties-graph.py
#
# @lc app=leetcode id=3493 lang=python3
#
# [3493] Properties Graph
#

# @lc code=start
class Solution:
    def numberOfComponents(self, properties: List[List[int]], k: int) -> int:
        n = len(properties)
        # Build adjacency list
        adj = [[] for _ in range(n)]
        def intersect(a, b):
            return len(set(a) & set(b))
        for i in range(n):
            for j in range(i+1, n):
                if intersect(properties[i], properties[j]) >= k:
                    adj[i].append(j)
                    adj[j].append(i)
        # Verify adjacency: each edge is undirected and intersection condition holds
        # (Omitted in code for efficiency, but should be considered in reasoning)
        visited = [False] * n
        def dfs(u):
            stack = [u]
            while stack:
                node = stack.pop()
                for v in adj[node]:
                    if not visited[v]:
                        visited[v] = True
                        stack.append(v)
        components = 0
        for i in range(n):
            if not visited[i]:
                visited[i] = True
                dfs(i)
                components += 1
        return components
# @lc code=end

File: 3493.properties-graph.java
#
# @lc app=leetcode id=3493 lang=java
#
# [3493] Properties Graph
#
# @lc code=start
class Solution {
    public int numberOfComponents(int[][] properties, int k) {
        int n = properties.length;
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                Set<Integer> setA = new HashSet<>();
                for (int x : properties[i]) setA.add(x);
                Set<Integer> setB = new HashSet<>();
                for (int x : properties[j]) setB.add(x);
                setA.retainAll(setB);
                if (setA.size() >= k) {
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }
        boolean[] visited = new boolean[n];
        int components = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(i, adj, visited);
                components++;
            }
        }
        return components;
    }
    private void dfs(int node, List<List<Integer>> adj, boolean[] visited) {
        visited[node] = true;
        for (int nei : adj.get(node)) {
            if (!visited[nei]) dfs(nei, adj, visited);
        }
    }
}
# @lc code=end